pkg/executor/sortexec/parallel_sort_spill_helper.go

```diff
@@ -0,0 +1,225 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sortexec
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/pingcap/tidb/pkg/types"
+	"github.com/pingcap/tidb/pkg/util"
+	"github.com/pingcap/tidb/pkg/util/chunk"
+	"github.com/pingcap/tidb/pkg/util/logutil"
+	"go.uber.org/zap"
+)
+
+type parallelSortSpillHelper struct {
+	cond             *sync.Cond
+	spillStatus      int
+	sortedRowsInDisk []*chunk.DataInDiskByChunks
+	sortExec         *SortExec
+
+	lessRowFunc   func(chunk.Row, chunk.Row) int
+	errOutputChan chan rowWithError
+	finishCh      chan struct{}
+
+	fieldTypes    []*types.FieldType
+	tmpSpillChunk *chunk.Chunk
+
+	bytesConsumed atomic.Int64
+	bytesLimit    atomic.Int64
+}
+
+func newParallelSortSpillHelper(sortExec *SortExec, fieldTypes []*types.FieldType, finishCh chan struct{}, lessRowFunc func(chunk.Row, chunk.Row) int, errOutputChan chan rowWithError) *parallelSortSpillHelper {
+	return &parallelSortSpillHelper{
+		cond:          sync.NewCond(new(sync.Mutex)),
+		spillStatus:   notSpilled,
+		sortExec:      sortExec,
+		lessRowFunc:   lessRowFunc,
+		errOutputChan: errOutputChan,
+		finishCh:      finishCh,
+		fieldTypes:    fieldTypes,
+		tmpSpillChunk: chunk.NewChunkWithCapacity(fieldTypes, spillChunkSize),
+	}
+}
+
+func (p *parallelSortSpillHelper) isNotSpilledNoLock() bool {
+	return p.spillStatus == notSpilled
+}
+
+func (p *parallelSortSpillHelper) isInSpillingNoLock() bool {
+	return p.spillStatus == inSpilling
+}
+
+func (p *parallelSortSpillHelper) isSpillNeeded() bool {
+	p.cond.L.Lock()
+	defer p.cond.L.Unlock()
+	return p.spillStatus == needSpill
+}
+
+func (p *parallelSortSpillHelper) isSpillTriggered() bool {
+	p.cond.L.Lock()
+	defer p.cond.L.Unlock()
+	return len(p.sortedRowsInDisk) > 0
+}
+
+func (p *parallelSortSpillHelper) setInSpilling() {
+	p.cond.L.Lock()
+	defer p.cond.L.Unlock()
+	p.spillStatus = inSpilling
+}
+
+func (p *parallelSortSpillHelper) setNeedSpillNoLock() {
+	p.spillStatus = needSpill
+}
+
+func (p *parallelSortSpillHelper) setNotSpilled() {
+	p.cond.L.Lock()
+	defer p.cond.L.Unlock()
+	p.spillStatus = notSpilled
+}
+
+func (p *parallelSortSpillHelper) spill() (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = util.GetRecoverError(r)
+		}
+	}()
+
+	select {
+	case <-p.finishCh:
+		return nil
+	default:
+	}
+
+	workerNum := len(p.sortExec.Parallel.workers)
+	workerWaiter := &sync.WaitGroup{}
+	workerWaiter.Add(workerNum)
+	sortedRowsIters := make([]*chunk.Iterator4Slice, workerNum)
+	for i := 0; i < workerNum; i++ {
+		go func(idx int) {
+			defer func() {
+				if r := recover(); r != nil {
+					processPanicAndLog(p.errOutputChan, r)
+				}
+				workerWaiter.Done()
+			}()
+
+			sortedRowsIters[idx] = chunk.NewIterator4Slice(nil)
+			sortedRowsIters[idx].Reset(p.sortExec.Parallel.workers[idx].sortLocalRows())
+			injectParallelSortRandomFail(200)
+		}(i)
+	}
```

Comment:
Is executing concurrently here necessary?

Reply:
> Is executing concurrently here necessary?

Yes, as each worker should sort their own rows.